internal/db/merge.go

```diff
@@ -0,0 +1,361 @@
+// Copyright 2024 Democratized Data Foundation
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package db
+
+import (
+	"container/list"
+	"context"
+
+	"github.com/ipfs/go-cid"
+	"github.com/ipld/go-ipld-prime/linking"
+	cidlink "github.com/ipld/go-ipld-prime/linking/cid"
+
+	"github.com/sourcenetwork/corelog"
+	"github.com/sourcenetwork/defradb/client"
+	"github.com/sourcenetwork/defradb/datastore"
+	"github.com/sourcenetwork/defradb/errors"
+	"github.com/sourcenetwork/defradb/events"
+	"github.com/sourcenetwork/defradb/internal/core"
+	coreblock "github.com/sourcenetwork/defradb/internal/core/block"
+	"github.com/sourcenetwork/defradb/internal/db/base"
+	"github.com/sourcenetwork/defradb/internal/merkle/clock"
+	merklecrdt "github.com/sourcenetwork/defradb/internal/merkle/crdt"
+	"github.com/sourcenetwork/immutable"
+)
+
+func (db *db) handleMerges(ctx context.Context, merges events.Subscription[events.DAGMerge]) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case merge, ok := <-merges:
+			if !ok {
+				return
+			}
+			go func() {
```

Comment:
> This is very resource greedy and inefficient
I very much disagree. Go routines that are non blocking are very much ok. It will exit safely on its own when it's done processing the merge. Not doing this would make our merge process single threaded and super slow in heavy usage environments.  HTTP servers do exactly this when receiving request. One go routine per request.

The problems with the leaking go routines that we are seeing are caused by them requiring a signal to exit. Those can be more problematic if not managed properly.

Reply:
> make our merge process single threaded

To clarify, I am very much not suggesting that we do this.

Reply:
> > This is very resource greedy and inefficient
> 
> I very much disagree. 

At the moment, our user-space requests operate on a single routine (roughly).  Each request can spawn a large number of child routines for stuff that does not take place in user-time.  This means that operations taking place in user time are going to be down-prioritised by the machine, when in fact they are very likely the most time sensitive.

> The problems with the leaking go routines that we are seeing are caused by them requiring a signal to exit.

This is one of several problems, all of which are IMO caused by a culturally relaxed attitude towards them.